### PR TITLE
[SPARK-5719]allow daemons to bind to specified host

### PR DESCRIPTION
**Pull request by WangTaoTheTonic**
https://issues.apache.org/jira/browse/SPARK-5719

Add four config items to specify the host that daemons bind to:

spark.master.host
spark.worker.host
spark.ui.host
spark.history.ui.host

### CODE REVIEW

**Comment by WangTaoTheTonic** [comment]
As now we have these configs, is it necessary to keep `SPARK_PUBLIC_DNS`? 

If yes, in which order should we place it?

File: core/src/main/scala/org/apache/spark/deploy/worker/WorkerArguments.scala [hunk 1]
```diff
@@ -23,46 +23,56 @@ import org.apache.spark.util.{IntParam, MemoryParam, Utils}
...
-  var host = Utils.localHostName()
-  var port = 0
-  var webUiPort = 8081
-  var cores = inferDefaultCores()
-  var memory = inferDefaultMemory()
+  var host: String = null
+  var port: Int = -1
+  var webUiPort: Int = -1
+  var cores: Int = -1
+  var memory: Int = -1
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/worker/WorkerArguments.scala]: you don't need to change the defaults here if you do the following down there:
```
port = sys.env.get(...)
  .orElse(conf.getOption(...))
  .getOrElse(port)
```

File: core/src/main/scala/org/apache/spark/deploy/worker/WorkerArguments.scala [hunk 2]
```diff
@@ -23,46 +23,56 @@ import org.apache.spark.util.{IntParam, MemoryParam, Utils}
...
+   * Load arguments from environment variables, Spark properties etc.
+   */
+  private def loadEnvironmentArguments(): Unit = {
+    host = Option(host)
+      .orElse(conf.getOption("spark.worker.host"))
+      .getOrElse(Utils.localHostName())
+    if (port < 0) {
+      port = conf.getOption("spark.worker.port").map(_.toInt)
+        .orElse(Option(conf.getenv("SPARK_WORKER_PORT")).map(_.toInt))
+        .getOrElse(0)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/worker/WorkerArguments.scala]: why not just `map(_.toInt)` after the `orElse`? Less duplicate code there

**Comment by andrewor14** [comment]
@WangTaoTheTonic what is the motivation for binding to a different host other than `0.0.0.0`? IIUC binding to `0.0.0.0` means you bind to all local interfaces. If we make this configurable then we're essentially allowing the user to bind to an address that is not part of this machine, which doesn't make sense.

**Comment by WangTaoTheTonic** [comment]
Assume that machines in our cluster have two network interfaces, one is used for management mainly while the other is used for business. Spark is the only one of many components and we just wanna access its Web UI via one of the network interfaces. That makes configurable ip address necessary.

Most configuration could be assigned with a wrong value, it will be ok as long as we give an error prompt when user configure it improperly.

File: core/src/main/scala/org/apache/spark/deploy/worker/WorkerArguments.scala (hunk 1 above)
Comment (WangTaoTheTonic) [inline core/src/main/scala/org/apache/spark/deploy/worker/WorkerArguments.scala]: How could we know whether `port` is already assigned with the value after `--port`(in the function `parse`)?

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/worker/WorkerArguments.scala]: ah I see

**Comment by andrewor14** [comment]
I'm still not sure if this is a use case we should support. Maybe others @JoshRosen @srowen have some thoughts.

**Comment by srowen** [comment]
@WangTaoTheTonic Isn't this what things like `SPARK_MASTER_IP` are for?

**Comment by JoshRosen** [comment]
@WangTaoTheTonic, to clarify the use-case, is the goal here to _prevent_ requests on certain network interfaces from being able to access Spark components?

This doesn't eliminate the need for SPARK_PUBLIC_DNS, which is used for controlling the advertised externally-routable hostname for things like web UI links, etc.

**Comment by andrewor14** [comment]
@WangTaoTheTonic let's close this patch since I don't think there is sufficient interest in this feature to warrant the additional complexity it introduces.

**Comment by pfxuan** [comment]
@WangTaoTheTonic , this could be a useful feature for further integration with other systems:

http://hortonworks.com/blog/multihoming-on-hadoop-yarn-clusters/
YARN: https://issues.apache.org/jira/browse/YARN-1994
Hadoop: https://issues.apache.org/jira/browse/HADOOP-8198
MapReduce: https://issues.apache.org/jira/browse/MAPREDUCE-4168
Tachyon: https://github.com/amplab/tachyon/pull/1195

**Comment by WangTaoTheTonic** [comment]
@JoshRosen @andrewor14 @srowen @pfxuan 
As the network related configurations need to be considered globally, I will close this for now.

**Pull request closed without merging**